pkg/controller/jaeger/jaeger_controller.go

```diff
@@ -222,3 +232,57 @@ func (r *ReconcileJaeger) apply(jaeger v1.Jaeger, str strategy.S) error {
 
 	return nil
 }
+
+//addFinalizer will add this attribute to the Jaeger CR
+func (r *ReconcileJaeger) addFinalizer(instance *v1.Jaeger) error {
+	if len(instance.GetFinalizers()) < 1 && instance.GetDeletionTimestamp() == nil {
+		log.Info("Adding Finalizer for the Jaeger")
+		instance.SetFinalizers([]string{finalizer})
+		// Update CR
+		err := r.client.Update(context.Background(), instance)
+		if err != nil {
+			log.Error(err, "Failed to update Jaeger with finalizer")
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileJaeger) clean(instance *v1.Jaeger) error {
+	opts := &client.ListOptions{}
```

Comment:
If you have thousands of deployments in the namespace, you'd get one thousand deployments as a result :-) 

See if you can filter that by the annotation (`sidecar.jaegertracing.io/inject`). Bonus points if you can filter by key=value (`sidecar.jaegertracing.io/inject=instance.Name`). This way, you get exactly the deployments you need to clean.

If you can do that, this PR would potentially be smaller.

Reply:
I didn't find a way to select or query by annotations, only with labels and fields :( That is why I did in this way.

Reply:
From here: https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/

> 
Labels can be used to select objects and to find collections of objects that satisfy certain conditions. In contrast, annotations are not used to identify and select objects.

Reply:
Could we add a label to the deployment when the sidecar is injected? If we can do that I can query for that label in the deletion logic.

Reply:
What do you think, @objectiser? I think I would change the current behavior, so that the annotation's value won't be touched. Instead, we add a label with the name of the instance the agent is talking to.

Reply:
Sounds fine to me - in which case we should only have a bool value for the inject annotation (rather than either bool or instance name) just to keep it simple. (Not sure if there would be migration issue?).

Reply:
As a last resort, we could define a "StringOrBool" type which would be able to parse both.